plugins/window-rules/window-rules.cpp

```diff
@@ -0,0 +1,234 @@
+#include <algorithm>
+#include <cfloat>
+#include <memory>
+#include <vector>
+
+#include <wayfire/plugin.hpp>
+#include <wayfire/view.hpp>
+#include <wayfire/view-access-interface.hpp>
+#include <wayfire/signal-definitions.hpp>
+#include <wayfire/view-transform.hpp>
+#include <wayfire/parser/rule_parser.hpp>
+#include <wayfire/lexer/lexer.hpp>
+#include <wayfire/variant.hpp>
+#include <wayfire/rule/lambda_rule.hpp>
+#include <wayfire/rule/rule.hpp>
+#include <wayfire/util/log.hpp>
+
+#include "lambda-rules-registration.hpp"
+#include "view-action-interface.hpp"
+
+class wayfire_window_rules_t : public wf::plugin_interface_t
+{
+  public:
+    void init() override;
+    void fini() override;
+    void apply(const std::string & signal, wf::signal_data_t *data);
+
+  private:
+    wf::lexer_t _lexer;
+
+    wf::signal_callback_t _created;
+    wf::signal_callback_t _maximized;
+    wf::signal_callback_t _unmaximized;
+    wf::signal_callback_t _minimized;
+    wf::signal_callback_t _fullscreened;
+
+    std::vector<std::shared_ptr<wf::rule_t>> _rules;
+
+    wf::view_access_interface_t _access_interface;
+    wf::view_action_interface_t _action_interface;
+
+    nonstd::observer_ptr<wf::lambda_rules_registrations_t> _lambda_registrations;
+};
+
+void wayfire_window_rules_t::init()
+{
+    // Get the lambda rules registrations.
+    _lambda_registrations = wf::lambda_rules_registrations_t::get_instance();
+
+    // TODO: Remove debug/test code.
```

Comment:
> A question: what is the purpose of the separated action interface? I thought its purpose is so that other plugins can also add their own action interfaces (not whole rules), but this doesn't seem the case.

It was split to get the ability to use the conditions separate from the action rules. You already do this in the matcher I think. The condition only needs the access interface where the rules (which act on the result of the condition check) need both interfaces.

I basically split this to build in more flexibility in the architecture with eye on re-usability of the rule/condition stuff. I would advise against merging these again.

Reply:
Ok.

Reply:
Test code removed